sympy/utilities/randtest.py

```diff
@@ -52,6 +57,56 @@ def verify_numerically(f, g, z=None, tol=1.0e-6, a=2, b=-1, c=3, d=1):
     return comp(z1, z2, tol)
 
 
+def test_at_rationals(f, g, z=None, attempts=5, a=-5, b=5, digits=3,
+        randomize=True):
+    """
+    Tests whether f and g yield exactly the same expression when symbol
+    z is replaced by a rational number. Does not attempt floating point
+    evaluation or any symbolic simplification, thus returns quickly.
+    Suitable for polynomials with rational coefficients because, for
+    example, ``x*(x+1) - x - x**2`` evaluates to zero when any rational
+    number is plugged in.
+
+    If z is None, all symbols will be tested. The test consists of n
+    substitutions of rational numbers, which are taken from interval
+    [a, b]. The parameter digits prescribes the approximate number of
+    digits in the denominator.
+
+    Examples
+    ========
+
+    >>> from sympy.abc import x
+    >>> from sympy.utilities.randtest import test_at_rationals
+    >>> test_at_rationals(x*(x + 1), x + x**2)
+    True
+    """
+    f, g, z = Tuple(f, g, z)
+    h = f - g
+    z = {z} if isinstance(z, Symbol) else h.free_symbols
+    if not z:
+        return h == 0
+    tol = 10**(-digits)
+    if randomize:   # use random rational numbers
+        for _ in range(attempts):
+            substitutions = {v: random_complex_number(a, 0, b, 0, rational=True,
+                tolerance=tol) for v in z}
+            if h.xreplace(substitutions) != 0:
+                return False
+    else:  # use cyclic shifts by golden ratio within [a, b]
+        counter = 0
+        golden_ratio = ((sqrt(5) + 1)/2).n(digits + 1)
```

Comment:
You can spell this `GoldenRatio` (as easy as `pi`). 

BTW, anything significant about this choice?

Reply:
Please see #14016 -- it seems like anything that can be proved by testing with Rationals should be easy to prove with `equals` -- but I suspect there is much of which I am unaware.

Reply:
The significance of golden ratio: cyclic shift (rotation) by this account create the most uniform aperiodic pattern possible. It has to do with its continued fraction expansion. I blogged about this at https://calculus7.org/2016/01/17/irrational-sunflowers/

Re: the issue. One situation that immediately comes to mind is having a product with one very complicated tem and another which is zero. Might algebraic expansion spend a lot of time on the complicated one, before noticing the other is zero?

Reply:
Very nice read on the sunflowers! Thanks.

Regarding algebraic expansion and rational numbers. I suspect that rational numbers will only resolve rational expressions. For non-rational expressions [as here](https://github.com/sympy/sympy/issues/5149) substituting in rational test values results in an expression that can't be computed with precision...and we can't guess in these cases. In some cases we simply run out of precision to prove that a number is not zero. So the rationals don't help. But if the expression is rational, we should be able to expand the expression symbolically and cancel terms and return a 0 if the expression truly is zero. Will it be quicker than substituting in a rational?

```
>>> r=lambda:random_complex_number(-1,-1,1,1, rational=True)
>>> eq=(1+x)**20 - ((1+x)**20).expand()
>>> t=time();eq.expand();time()-t
0
0.188018798828125
>>> t=time();eq.subs(x,re(r()));time()-t
0
0.1770174503326416

If a number has rational real and complex parts, the result has to be separated into
real and imaginary parts to see the zero -- attempting to compute it without doing
so leads to a precisionless 0

>>> eq.subs(x,r()).n()
0.e-130 - 0.e-130*I
>>> t=time();eq.subs(x,r()).as_real_imag();time()-t
(0, 0)
0.4750478267669678
```

Can you think of a complicated expression where rational testing will work? Is it necessary to use imaginary parts when testing? I tried a higher power expression as above and the expansion out-performs substitution even more:

```
>>> eq=(1+x)**50 - ((1+x)**50).expand()
>>> t=time();eq.subs(x,re(r()));time()-t
0
0.31203126907348633
>>> t=time();eq.expand();time()-t
0
0.014001131057739258
```

BTW, that expression in #5149 could probably be made to have a radius of any size in which the result would be zero. I don't remember now where I got that expression. It seems like it must have been concocted to be that way with `x**2 + y**2 = 1`. Maybe you wouldn't run into such expressions in a single variable.